core/src/main/java/com/linecorp/armeria/client/RefreshingAddressResolver.java

```diff
@@ -0,0 +1,369 @@
+                records.forEach(ReferenceCountUtil::safeRelease);
+            }
+
+            if (inetAddress == null) {
+                result.completeExceptionally(
+                        new UnknownHostException("failed to receive DNS records for " + hostname));
+                return;
+            }
+            final CacheEntry entry = new CacheEntry(inetAddress, ttlMillis, questions);
+            result.complete(entry);
+            entry.scheduleCacheUpdate(ttlMillis);
+        });
+    }
+
+    @Override
+    protected void doResolveAll(InetSocketAddress unresolvedAddress, Promise<List<InetSocketAddress>> promise)
+            throws Exception {
+        throw new UnsupportedOperationException();
+    }
+
+    /**
+     * Closes all the resources allocated and used by this resolver.
+     *
+     * <p>Please note that this method does not clear the {@link CacheEntry} because the {@link #cache} is
+     * created by {@link RefreshingAddressResolverGroup} and shared across from all
+     * {@link RefreshingAddressResolver}s. {@link CacheEntry} is cleared when
+     * {@link RefreshingAddressResolverGroup#close()} is called.
+     */
+    @Override
+    public void close() {
+        resolverClosed = true;
+        resolver.close();
+    }
+
+    final class CacheEntry implements Runnable {
+
+        private final InetAddress address;
+        private final long ttlMillis;
+        private final List<DnsQuestion> questions;
+
+        /**
+         * No need to be volatile because updated only by the {@link #executor()}.
+         */
+        private int numAttemptsSoFar = 1;
+
+        private volatile boolean servedFromCache;
+
+        @Nullable
+        volatile ScheduledFuture<?> cacheUpdatingScheduledFuture;
+
+        CacheEntry(InetAddress address, long ttlMillis, List<DnsQuestion> questions) {
+            this.address = address;
+            this.ttlMillis = ttlMillis;
+            this.questions = questions;
+        }
+
+        void servedFromCache() {
+            servedFromCache = true;
+        }
+
+        InetAddress address() {
+            return address;
+        }
+
+        long ttlMillis() {
+            return ttlMillis;
+        }
+
+        void scheduleCacheUpdate(long nextDelayMillis) {
+            if (resolverClosed) {
+                return;
+            }
+            final ScheduledFuture<?> oldFuture = futureUpdater.get(this);
+            if (oldFuture == closed) {
+                return;
+            }
+
+            final ScheduledFuture<?> newFuture =
+                    executor().schedule(this, nextDelayMillis, TimeUnit.MILLISECONDS);
+            if (!futureUpdater.compareAndSet(this, oldFuture, newFuture)) {
+                // clear() is called and the future is set to closed future. So we just cancel the newFuture
+                // we just made.
+                newFuture.cancel(true);
+            }
+        }
+
+        void clear() {
+            assert resolverClosed;
```

Comment:
I think clear is only called after closing the resolver and I guess this assert agrees. Since we don't reschedule or process a response if the resolver is closed, I still feel like we shouldn't need the updater for the future. Closing the client factory to close the resolvers is rare, so if we want to cancel the future I think we can just schedule cancellation on executor () instead of making the standard case of updating more complicated by the updater. I might not even worry about cancel since even if the future runs it will finish in microseconds, doing nothing since the resolver is closed.

Maybe the summary is that since we already have one volatile boolean (resolverClosed) which we check everywhere, we probably don't need another one.

Reply:
There's a chance that the future is not canceled after the resolver is closed, but yes, we are checking `resolverClosed` and it won't be a big deal because it would finish in microseconds.
Let me change that. Thanks!